source/core/StringUtilities.cpp

```diff
@@ -1132,6 +1132,64 @@ DecodedSubString::DecodedSubString(const SubString &s, bool decode, bool alwaysA
     Init(s, decode, alwaysAlloc);
 }
 
+const char* HEX_VALUES = "0123456789ABCDEF";
+
+void EncodedSubString::Init(const SubString &s, bool encode, bool alwaysAlloc) {
+    const Utf8Char *cp = s.Begin();
+    IntIndex len = s.Length();
+    if (_encodedStr && _encodedStr != _buffer)
+        delete[] _encodedStr;
+    if (encode) {
+        while (cp != s.End()) {
+            if (cp == s.Begin() && !SubString::IsLetterChar(*cp)) {
+                len += 2;
+                 ++cp;
+                continue;
+            }
+            if (NeedsEncoding(*cp))
+                len += 2;
+            ++cp;
+        }
+    }
+    if (alwaysAlloc || len >= kMaxInlineEncodedSize)
+        _encodedStr = new Utf8Char[len+1];
+    else
+        _encodedStr = _buffer;
+    if (encode) {
+        SubString ss = s;
+        Utf8Char c;
+        Int32 value;
+        Utf8Char *pDest = _encodedStr;
+
+        if (ss.ReadRawChar(&c)) {
+            if (SubString::IsLetterChar(c)) {
+                *pDest++ = c;
+            } else {
+                *pDest++ = '%';
+                *pDest++ = (Utf8Char)(HEX_VALUES[(0xF0 & c) >> 4]);
```

Comment:
If we are not concerned about performance, consider adding a utility function to make the code clearer. Something along GetHexadecimalRepresentationForUtf8Char, that will take pDest and c as parameters and basically do these 3 lines of code.